### PR TITLE
[Core] Cleaning up and adding templates to some methods on GeometryUtilities

### PR DESCRIPTION
**Pull request by loumalouomega**
This cleans the GeometryUtilities and adds template arguments to some functions (so bounded matrices can be used)

### CODE REVIEW

**Comment by loumalouomega** [comment]
Ping

**Comment by pablobecker** [comment]
Lots of changes here (high possiblity to add bugs?) 
do these functions have tests?

**Comment by loumalouomega** [comment]
Many of them not, I will add it

In any case the changes are adapting to style guide and adding missing const (and deprecating duplicated methods)

**Comment by loumalouomega** [comment]
Still some tests missing, on it. I fixed some things that weren't working even in the master

**Comment by loumalouomega** [comment]
@pablobecker Everything tested

**Comment by loumalouomega** [comment]
@roigcarlo can you check why this test fails in Travis?, in my machine both FullDebug and Release works fine

File: kratos/tests/cpp_tests/utilities/test_geometry_utils.cpp
Comment (loumalouomega) [inline kratos/tests/cpp_tests/utilities/test_geometry_utils.cpp]: @roigcarlo Is this test
```suggestion
//        KRATOS_CHECK_EQUAL(intersections, 1);
```

**Comment by pooyan-dadvand** [comment]
I don't know how we arrive to this point having so many copy of the determinant and Jacobians over all Kratos... @RiccardoRossi this is our fault

Meanwhile too many changes in a basic tool

**Comment by loumalouomega** [comment]
> I don't know how we arrive to this point having so many copy of the determinant and Jacobians over all Kratos... @RiccardoRossi this is our fault
> 
> Meanwhile too many changes in a basic tool

I added tests to everything (in fact we didn't have tests previously)

File: kratos/utilities/geometry_utilities.h
Comment (philbucher) [inline kratos/utilities/geometry_utilities.h:699]: this will not compile with AMatrix, same issue that was already discussed in the previous days

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:699]: True

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:699]: I "fixed" on https://github.com/KratosMultiphysics/Kratos/pull/4144/commits/6d86cb2446f895d464fd300ca5a4c8b467389219 sincerely is difficult to tackle that

**Comment by pablobecker** [comment]
@KratosMultiphysics/implementation-committee has reached consensus. These changes are correct and we are waiting for the final decision from the @KratosMultiphysics/technical-committee

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:181]: same

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h]: noalias

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h]: noalias

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:114]: do the tet and triangle implement this efficiently, or do they use the generic version?

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:316]: is this large change only cosmetic?

File: kratos/utilities/geometry_utilities.h
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:479]: only cosmetic?

File: kratos/utilities/geometry_utilities.h
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:181]: Same as what?

File: kratos/utilities/geometry_utilities.h
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:479]: Yes

File: kratos/utilities/geometry_utilities.h
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:316]: Yes

File: kratos/utilities/geometry_utilities.h
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:114]: The efficient one:

https://github.com/KratosMultiphysics/Kratos/blob/6b06365736c430f72eaff3ac52a0033ca39242e0/kratos/geometries/tetrahedra_3d_4.h#L404

**Review by RiccardoRossi — COMMENTED** [review]
this changes look ok to me.
@rubenzorrilla you making use of some of the functions employed here. please do the approving.

**Comment by loumalouomega** [comment]
Ping

**Comment by rubenzorrilla** [comment]
As far as I can see, this only includes cosmetic changes. Am I wrong?

**Comment by loumalouomega** [comment]
Yep

**Review by rubenzorrilla — APPROVED** [review]

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**
